kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt

```diff
@@ -53,7 +54,7 @@ class BigIntegerGen(maxNumBits: Int) : Gen<BigInteger> {
  * The [Int] generator example should return a random int
  * from across the entire integer range.
  */
-interface Gen<T> {
+@higherkind interface Gen<T> : GenOf<T> {
```

Comment:
Where is GenOf defined?

Reply:
It's created by processing `@higherkind`:

```
@higherkind data class ListK<A>(val list: List<A>): ListKOf<A>

// Generates the following code:
//
// class ForListK private constructor() { companion object {} }
// typealias ListKOf<A> = Kind<ForListK, A>
// fun ListKOf<A>.fix() = this as ListK<A>
```

Reply:
So by adding @higherkind to `Foo`, `FooOf<A>` is generated, and is a type alias to Kind<ForFoo, A>, where `ForFoo` is another automatically generated class. Why does the type alias require ForFoo? Is this to extract the type parameter basically?

Reply:
This is the full explanation:
https://arrow-kt.io/docs/patterns/glossary/#higher-kinds

The type alias is a nicety to make it easier to read and write extension functions for one datatype. 

```
fun <A, B, C> ListKOf<A>.doThing(another: ListKOf<B>): ListK<C>
```
Parameters are contravariant so you always take a ListK, and returns are covariant so you can return the specific class. This makes the extfuns work pretty much for everywhere, and their returns don't require `fix()` anymore.

```
val a: Kind<ForOption, ListK<User>> = userOptions.sequence(Option.applicative())
val b: Option<Tuple2<Int, ListK<User>> = a.tupleLeft(2)
```

Reply:
The key here is that Kotlin allows extfuns to be specialized for a generic, so `Kind<ForOption, A>` can have a different implementation that `Kind<ForListK, A>` even if they're defined on the same file.

Reply:
It's coming together now in my tiny brain.
You're doing all this machinery basically because you can't do this:

```kotlin
interface Functor<F<A>> {
  fun <B> map(f: (A) -> B): F<B>
}
```

Reply:
```
interface Functor<F<_>> {
  fun <A, B> F<A>.map(f: (A) -> B): F<B>
}
```

Reply:
Ok yes, because you don't care about A when you're implementing ListFunctor or whatever.